tests/performance/density.go

```diff
@@ -0,0 +1,471 @@
+	var (
+		queue   = make(chan map[e2emetrics.MetricType]e2emetrics.VMMetricData)
+		e2eData = make(map[e2emetrics.MetricType]e2emetrics.Slice)
+	)
+
+	By("Creating a batch of vms")
+	firstCreate := time.Now()
+	vmList := createBatchVMWithRateControl(virtClient, testArg, queue)
+	By("Waiting for all VMIs become ready...")
+	waitResults(queue, &e2eData, testArg, e2emetrics.Running, testArg.VMBatchStartupLimit)
+
+	lastRunning := time.Now()
+	batchStartupLag := lastRunning.Sub(firstCreate) / time.Second
+	fmt.Printf("total batch VMI startup latency %d", batchStartupLag)
+
+	// If the cluster has Prometheus running and it is accessible
+	if resColl != nil {
+		By("Collect VirtualMachineInstance Metrics from Prometheus")
+		queue = make(chan map[e2emetrics.MetricType]e2emetrics.VMMetricData)
+		for _, vmi := range vmList {
+			go collectVMIMetrics(virtClient, queue, resColl, vmi)
+		}
+		By("Wait CPU and MEM metrics be collected to all VMIs")
+		waitResults(queue, &e2eData, testArg, e2emetrics.VMIPodMemory, testArg.VMBatchStartupLimit)
+	}
+
+	By("Deleting all VirtualMachineInstance")
+	firstDelete := time.Now()
+	queue = make(chan map[e2emetrics.MetricType]e2emetrics.VMMetricData)
+	for _, vmi := range vmList {
+		go waitVMITerminate(virtClient, queue, vmi, testArg.VMDeletionLimit*time.Second)
+	}
+	By("Wait all VirtualMachineInstance Disappear")
+	waitResults(queue, &e2eData, testArg, e2emetrics.Terminated, testArg.VMDeletionLimit)
+
+	By("Collect Control Plane Metrics from Prometheus")
+	if resColl != nil {
+		CollectControlPlaneMetrics(&e2eData, resColl)
+	}
+
+	lastRunning = time.Now()
+	batchTerminationLag := lastRunning.Sub(firstDelete) / time.Second
+	fmt.Printf("total batch VMI deletetion latency %d", batchStartupLag)
+
+	return batchStartupLag, batchTerminationLag, e2eData
+}
+
+func waitResults(queue chan map[e2emetrics.MetricType]e2emetrics.VMMetricData, e2eData *map[e2emetrics.MetricType]e2emetrics.Slice,
+	testArg DensityTest, stopType e2emetrics.MetricType, timout time.Duration) {
+	timeoutq := time.After(timout * time.Second)
+	var err error
+startupChannel:
+	for {
+		select {
+		case <-timeoutq: // timed out
+			err = fmt.Errorf("failed for %s of %v VMIs from %v within the expected interval %v",
+				stopType.String(),
+				testArg.NumVMs-len((*e2eData)[e2emetrics.Running]),
+				testArg.NumVMs, testArg.VMBatchStartupLimit*time.Second)
+			break startupChannel
+
+		case e2eD := <-queue:
+			for metric, val := range e2eD {
+				(*e2eData)[metric] = append((*e2eData)[metric], val)
+			}
+
+			if len((*e2eData)[stopType]) == testArg.NumVMs {
+				err = nil
+				break startupChannel
+			}
+		}
+	}
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func attemptToLoginToVM(virtClient kubecli.KubevirtClient, testArg DensityTest, vmi *apiv1.VirtualMachineInstance) *apiv1.VirtualMachineInstance {
```

Comment:
This part I am not sure if I should keep in the PR.

Right now, I agreed with @rmohr that the test in our CI/CD system will not boot the VMs to be able to scale it and put more pressure in the control plane. The creation time is the time interval from submission to Running Phase.

However, I made the test as flexible as possible that someone could easily change the VM creation time from submission up to when we can ping the VM. To be able to test the whole VM creation pipeline, and make sure that all components are properly working (e.g. the network)...

Reply:
I spoke with @davidvossel previously about an idea I had that's similar.  I'll share the WIP design document at the next SIG-scale and we should discuss this idea as well.

Reply:
Sure! Thanks!
Btw, we could start this performance test with the approach that I am suggesting in this PR, and extend it in the future. wdyt?